Context Menu: New elements should be selected after duplication

samwhale opened the pull request:
## Context

Issue found during Bug Bash 1.16.0.

## Summary

Duplicating one or more elements will select the newly created element(s).

## Relevant Technical Choices

<!-- Please describe your changes. -->

## User-facing changes

|Multiple|Single|From Layer Panel|
|--|--|--|
|![duplicateelements](https://user-images.githubusercontent.com/22185279/149976723-b0dfa725-9982-4556-984f-8c6ebd186e0f.gif)|![layerduplicate](https://user-images.githubusercontent.com/22185279/149976997-449ca5d9-ede7-4bd0-9248-0b8365cb320b.gif)|![duplicateelement](https://user-images.githubusercontent.com/22185279/149977020-c9328577-7ed0-4869-87ac-9c3760532eda.gif)|

## Testing Instructions

<!-- ignore-task-list-start -->
- [ ] This is a non-user-facing change and requires no QA
<!-- ignore-task-list-end -->

You will need to test that duplicating elements still works as expected:
- duplicates all animations applied to element
- places newly duplicated element in a different spot

New functionality:
- the newly created element will be selected after duplication.
- clicking the duplication button in the layer panel should select the new element on creation.

This PR can be tested by following these steps:

1. Add element(s) to canvas. Add styles + animations to your heart's content
2. Select element(s)
3. Duplicate
4. Verify that new elements are selected and the old elements are not
5. Verify that elements were duplicated correctly


## Reviews

### Does this PR have a security-related impact?

no

### Does this PR change what data or activity we track or use?

no

### Does this PR have a legal-related impact?

no

## Checklist

<!-- Check these after PR creation -->

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

<!--
Please reference the issue(s) this PR addresses.
No URLs, just the issue numbers.
Use "Fixes #123" if it fixes an issue.

NOTE: One reference per line!

Example:

Fixes #123
Partially addresses #456
See #789
-->

Fixes #10179

samwhale committed:
Refactor duplicate elements by id
* Select all new elements created
* duplicate multiple elements at once

BrittanyIRL reviewed (approved):
🤝  🎸 🚀  

Works as described

miina commented on packages/story-editor/src/app/story/useStoryReducer/actions.js:
```diff
@@ -119,12 +119,12 @@ const updateElementById =
-  ({ elementId }) =>
+  ({ elementIds }) =>
     dispatch({
       type: types.DUPLICATE_ELEMENT_BY_ID,
```
Let's change this as well then everywhere for consistency:

```suggestion
      type: types.DUPLICATE_ELEMENTS_BY_ID,
```

miina commented on packages/story-editor/src/app/story/useStoryReducer/reducers/duplicateElementsById.js:
```diff
@@ -0,0 +1,91 @@
+
+    return;
+  });
+
```
Could we here just `return state` if `!newSelection.length`? Would avoid creating a new history entry if there are no  elements duplicated for some reason.

samwhale replied:
Ah great catch!

miina commented on packages/story-editor/src/app/story/useStoryReducer/reducers/setSelectedElements.js:
```diff
@@ -26,7 +26,7 @@ import { intersect } from './utils';
  * Set selected elements to the given list of ids.
  *
- * If given list is not a list, do nothing.
+ * If given `elementIds` are not a list, do nothing.
```
```suggestion
 * If given `elementIds` is not a list, do nothing.
```

miina commented on packages/story-editor/src/app/story/useStoryReducer/test/duplicateElementByIds.js:
```diff
@@ -169,15 +176,81 @@ describe('duplicateElementById', () => {
+    expect(initialElement.y).not.toBe(newElement.y);
+  });
+
+  it('duplicates more than one element but not the background', () => {
```
Maybe let's add one with incorrect element IDs as well and verify the initial state doesn't change.

miina reviewed (commented):
Nice work! Left a few nits.

samwhale committed:
Update packages/story-editor/src/app/story/useStoryReducer/reducers/setSelectedElements.js

Co-authored-by: Miina <miina.sikk@gmail.com>

samwhale committed:
PR Fixes
- Update names to be plural
- Add test to make sure wrong ids don't update state
- Return early if no elements to update

samwhale committed:
Add another test: multiple elements but one with wrong id

maxyinger reviewed (approved):
Code looks and works great!

Just want to make sure it's also intentional for the layers panel to select the new element when you click the duplicate icon on hover?

I also think one of the main reasons why this action was created in the layers panel vs what already exists in the codebase was to not have to select elements when they're duplicated. Now that it is doing that, we may be able to unify some implementations in the app. i.e.

https://github.com/google/web-stories-wp/blob/67d010eec4efe7e9797362f0610ecf55e24b8338/packages/story-editor/src/app/canvas/useCanvasKeys.js#L233-L262

samwhale committed:
Use the existing reducer fn duplicateElementsByIds in layer panel

samwhale committed:
Fix test

miina commented on packages/story-editor/src/app/story/useStoryReducer/test/duplicateElementByIds.js:
```diff
@@ -0,0 +1,346 @@
+/*
```
Nit: The filename should probably be `duplicateElementsById`.

miina commented on packages/story-editor/src/app/story/useStoryReducer/reducers/duplicateElementsById.js:
```diff
@@ -0,0 +1,96 @@
+    newSelection.push(element.id);
+
+    newPage.animations = [...(newPage.animations || []), ...elementAnimations];
+    newPage.elements = [
```
One thing that happens like this is that the duplicated layer is added right next to the origin element, e.g. if I duplicate 3 elements:
- IMAGE 1
- IMAGE 2
- IMAGE 3

it becomes
- IMAGE 1
- IMAGE 1 dupe
- IMAGE 2
- IMAGE 2 dupe
- IMAGE 3
- IMAGE 3 dupe

instead of 
- IMAGE 1
- IMAGE 2
- IMAGE 3
- IMAGE 1 dupe
- IMAGE 2 dupe
- IMAGE 3 dupe

Is this intended? I was thinking that all the dupes would go to the end of the list of the layers as new elements, however, not really sure what's the standard behaviour in this case for layers. Thoughts?

samwhale replied:
Good question. I don't have context as to why that happens, but this is what existed before this ticket:

https://github.com/GoogleForCreators/web-stories-wp/blob/main/packages/story-editor/src/app/story/useStoryReducer/reducers/duplicateElementById.js#L47-L52

@littlemilkstudio do you have more context here?

BrittanyIRL replied:
This should be a question for UX not eng - @agingoldseco  do you have a quick second to weigh in on this?

agingoldseco replied:
Can't say I have a good gauge of "standard" behavior here, but I think it make sense for copy/paste and duplicate to act differently. I think copy/paste the way it is now makes sense, where the new layers appear in the layers panel op top of everything else. And Duplicate also makes sense as you might want a duplicated element to appear on the same "level" as its origin element. But Morten's comment in bug bash is right—in my experience I want to interact with the duplicated elements first after duplication, so they should be selected rather than the origin elements. Just checked, and that's actually what Figma does too! 

Here's a video—see what happens in the left layers column when I copy the circles (first) and then duplicate them using option+drag:

https://user-images.githubusercontent.com/90645922/150835731-de1f427f-aea4-48b9-b2f4-b4eb9b128475.mov

samwhale replied:
@agingoldseco this pr leaves the newly created elements selected 👍 

Seems like figma also places the newly created elements right next to their parent elements. For now I'll leave the behavior as is, but it will be an easy change if in the future we'd like that to act the same as copy/paste.

samwhale committed:
Merge branch 'main' into feature/10179-select-duplicate-elements

samwhale committed:
Rename test file

samwhale committed:
Merge branch 'main' into feature/10179-select-duplicate-elements

samwhale committed:
Merge branch 'main' into feature/10179-select-duplicate-elements
